php/class-responsive-breakpoints.php

```diff
@@ -0,0 +1,336 @@
+				),
+				array(
+					'width'  => '$currWidth',
+					'height' => '$currHeight',
+					'crop'   => 'fill',
+				),
+				array(
+					'fetch_format' => 'auto',
+					'quality'      => 'auto',
+				),
+			),
+			'vectorize'   => array(
+				array(
+					'effect'       => 'vectorize:3:0.1',
+					'fetch_format' => 'svg',
+				),
+			),
+			'blur'        => array(
+				array(
+					'effect'       => 'blur:2000',
+					'quality'      => 1,
+					'fetch_format' => 'auto',
+				),
+			),
+			'pixelate'    => array(
+				array(
+					'effect'       => 'pixelate',
+					'quality'      => 1,
+					'fetch_format' => 'auto',
+				),
+			),
+		);
+
+		return $transformations[ $placeholder ];
+	}
+
+	/**
+	 * Add features to a tag element set.
+	 *
+	 * @param array  $tag_element   The tag element set.
+	 * @param int    $attachment_id The attachment id.
+	 * @param string $original_tag  The original html tag.
+	 *
+	 * @return array
+	 */
+	public function add_features( $tag_element, $attachment_id, $original_tag ) {
+
+		$settings = $this->settings->get_value();
+
+		$meta   = wp_get_attachment_metadata( $attachment_id, true );
+		$format = $this->media->get_settings()->get_value( 'image_format' );
+
+		$src                       = $tag_element['atts']['src'];
+		$transformations           = $this->media->get_transformations_from_string( $src );
+		$placehold_transformations = $transformations;
+		$original_string           = Api::generate_transformation_string( $transformations );
+		$breakpoints               = $this->media->get_settings()->get_value( 'enable_breakpoints' );
+		if ( 'on' === $breakpoints ) {
+			// Check if first is a size.
+			if ( isset( $transformations[0] ) && isset( $transformations[0]['width'] ) || isset( $transformations[0]['height'] ) ) {
+				// remove the size.
+				array_shift( $transformations );
+			}
+			$size_str                        = '--size--/' . Api::generate_transformation_string( $transformations );
+			$data_url                        = str_replace( $original_string, $size_str, $src );
+			$tag_element['atts']['data-src'] = $data_url;
+			if ( isset( $tag_element['atts']['srcset'] ) ) {
+				unset( $tag_element['atts']['srcset'], $tag_element['atts']['sizes'] );
+			}
+		}
+
+		// placeholder.
+		if ( 'off' !== $settings['lazy_placeholder'] ) {
+			// Remove the optimize (last) transformation.
+			array_pop( $placehold_transformations );
+			$placehold_transformations = array_merge( $placehold_transformations, $this->get_placeholder_transformations( $settings['lazy_placeholder'] ) );
+			$palcehold_str             = Api::generate_transformation_string( $placehold_transformations );
+			$placeholder               = str_replace( $original_string, $palcehold_str, $src );
+
+			$tag_element['atts']['src'] = $placeholder;
+			if ( ! empty( $settings['lazy_preloader'] ) ) {
+				$tag_element['atts']['data-placeholder'] = $placeholder;
+			}
+		}
+
+		if ( ! empty( $settings['lazy_preloader'] ) ) {
```

Comment:
@DavidCramer 

This is a reminder for introducing the controls discussed for the preloader:
+ Predominant color;
+ Color picker;
+ Fade animation;

Also the preloader should always be there.